crates/shared/src/price_estimation/native/coingecko.rs

```diff
@@ -0,0 +1,144 @@
+use {
+    super::{NativePriceEstimateResult, NativePriceEstimating},
+    crate::price_estimation::PriceEstimationError,
+    anyhow::{anyhow, Result},
+    futures::{future::BoxFuture, FutureExt},
+    primitive_types::H160,
+    reqwest::{header::AUTHORIZATION, Client, StatusCode},
+    serde::Deserialize,
+    std::collections::HashMap,
+    url::Url,
+};
+
+#[derive(Debug, Deserialize)]
+struct Response(HashMap<H160, Price>);
+
+#[derive(Debug, Deserialize)]
+struct Price {
+    eth: f64,
+}
+
+type Token = H160;
+
+pub struct CoinGecko {
+    client: Client,
+    base_url: Url,
+    api_key: Option<String>,
+    chain: String,
+}
+
+impl CoinGecko {
+    pub fn new(
+        client: Client,
+        base_url: Url,
+        api_key: Option<String>,
+        chain_id: u64,
+    ) -> Result<Self> {
+        let chain = match chain_id {
+            1 => "ethereum".to_string(),
+            100 => "xdai".to_string(),
+            42161 => "arbitrum-one".to_string(),
+            n => anyhow::bail!("unsupported network {n}"),
+        };
+        Ok(Self {
+            client,
+            base_url,
+            api_key,
+            chain,
+        })
+    }
+}
+
+impl NativePriceEstimating for CoinGecko {
+    fn estimate_native_price(&self, token: Token) -> BoxFuture<'_, NativePriceEstimateResult> {
+        async move {
+            let url = format!(
+                "{}/{}?contract_addresses={token:#x}&vs_currencies=eth",
+                self.base_url, self.chain
+            );
+            let mut builder = self.client.get(&url);
+            if let Some(ref api_key) = self.api_key {
+                builder = builder.header(AUTHORIZATION, api_key)
```

Comment:
Shouldn't we use [x_cg_pro_api_key](https://support.coingecko.com/hc/en-us/articles/4538879945625-API-key-can-I-store-the-key-in-my-headers) header instead?

Reply:
really good  catch @squadgazzz . Thank you.